### DISCUSSION TITLE
How would you support custom allocators?

### DISCUSSION THREAD

**Discussion by johnrrowe** [hylo-lang/hylo#1347] [Q&A]
This is a similar question to https://github.com/orgs/hylo-lang/discussions/736 but specifically about custom allocators. The general pattern in languages like C++ and Rust is to store a pointer/reference to the allocator, but that breaks MVS. I also don't think it always makes sense for a collection to own its allocator. For example, you might want to use a bump allocator to pack a bunch of heterogeneous objects together in memory. Is there an ergonomic solution to this that doesn't sacrifice MVS? I really like the design of Hylo, so I'm hoping there is. An obvious but undesirable solution would be an array of union objects, but that has multiple inefficiencies.

**Comment by dabrahams** [comment]
Thanks for asking!  

Except for global threadsafe shared resources like the system's malloc allocator, an allocator is a container,  a whole, and the things allocated within it are its parts. They can be projected out of the allocator.  That's the only way to maintain MVS.  It may not even make sense to think of the allocator as an allocator; you might think of it as a heterogeneous container.  The exact way that memory is acquired by the allocator and apportioned to the allocator's parts is of course dependent on the allocation strategy (arrays of raw bytes can work, with some bookkeeping).  The illusion presented by C++ that there is no whole-part relationship between the allocator and things stored in it cannot be upheld, so the ergonomics _will_ be different.  And that's about as much as I know for sure at this point.

Whether it's reasonable for  vocabulary collection types to be able to expose borrowed memory, or whether you need a separate family of collection types that can be projected out of allocators, is an open question.  Implicits (implicit function arguments) may have a role to play in keeping usage ergonomic also; they can help hide the passing-around of context.    But knowing more is going to take some experimentation, so I encourage you to dig into the problems yourself and see what you come up with.

**Reply by johnrrowe** [reply]
Thanks for your response. What you said makes sense, and I think I agree that the usual idea of an allocator implies a "references" relationship, not a whole-part. I haven't taken the time to use Hylo myself, but I definitely plan to explore these ideas further. The most straightforward translation of an allocator API I thought of looks like this (using Rust because I'm not very familiar with Hylo syntax). An allocator would be a heterogeneous container that gives access to its parts through a handle, where `project` would be a subscript in Hylo.

```rust
trait Allocator {
    // This probably needs to be opaque to users. Wouldn't want the internals of it to be modifiable
    type Handle<T>;

    fn alloc<T, F: FnOnce() -> T>(&mut self, construct: F) -> Self::Handle<T>;

    // SAFETY: Handle must have been allocated by the same instance. Value being projected must still be valid.
    unsafe fn project<'s, T>(&'s self, h: &Self::Handle<T>) -> &'s T;
    
    // SAFETY: Handle must have been allocated by the same instance. Handle must not have been copied
    // otherwise potential double free/use after free. Nothing is forcing the handle to be freed, so potential leaks.  
    unsafe fn free<T>(&mut self, h: Self::Handle<T>);
}
```

 I don't like that this API is inherently unsafe. Some improvements that I think are theoretically possible:
1. Require Handle to be non-copyable (not sure if that's expressible in Rust traits). I'm not sure if Hylo has support for "true" linear types in the sense that the compiler can force you to sink a value for "final destruction". Not sure what destructors look like in Hylo, if they exist, but I don't think that would be viable because Handle's destructor would need access to the allocator.
2. A way for the compiler to associate a Handle to a particular instance of an allocator. I came across an interesting Rust proposal based on static path-dependent types, which I think is inspired by Scala. I've never used them, so I don't fully understand how they work. But I think this API needs something like this to be safe. [Here's the paper](https://cfallin.org/pubs/ecoop2020_defborrow.pdf) | [Here's a video by the author](https://www.youtube.com/watch?v=k7mPB5u6mKI&ab_channel=ACMSIGPLAN)

All these hoops feel like I'm trying to reinvent reference semantics with extra steps, but at least things are more explicit than they would be with first class references, i.e. no spooky action. But to some extent, I think this is kind of required in MVS?

I think statically enforced path-dependent types are an interesting concept, but I'm not sure if Hylo would find their complexity/ergonomics acceptable. I could imagine them being as complicated as Rust's lifetime annotations. However, I think they're worth considering because they would be useful beyond allocators. Anytime you have a context that produces handles that are tied to a specific instance of said context, I think you could use path-dependent types to ensure that at compile time. For example, when using the adjacency list representation of a graph, you could ensure your node indices are used with the correct graph.

Edit: Looks like people have actually implemented a similar idea in Rust by encoding things with lifetime parameters [Crate](https://docs.rs/ghost-cell/latest/ghost_cell/) | [Video](https://www.youtube.com/watch?v=jIbubw86p0M&ab_channel=ACMSIGPLAN)

**Reply by kyouko-taiga** [reply]
> Require Handle to be non-copyable

Hylo has a true linear type system so you can make the handles non-copyable. You can also require that the only type knowing how to destroy a handle is the allocator. That's because a "destructor" is simply a method that sinks an object and can consume its contents.

You would need to gather and collect all handles before you destroy the allocator itself, though.

> A way for the compiler to associate a handle with a particular instance of the allocator

We would need Scala-like path-dependent types for that, indeed. However, we have concluded they didn't justify the complexity they bring to the language the last time we discussed them.

Having handles coming from multiple allocators of the same type in scope sounds like a very particular setup, though. I haven't personally encountered realistic examples of the kind. Further I became used to deal with positions into collections that do not "know" the collection instance from which they come (and neither does the compiler). That is essentially the same setup as the one you describe with allocators, except that the "handles" can be destroyed anywhere.

I think that in general we should strive for situations where the positions/handles of type T.X we use in one scope all come from a single instance of a container/allocator of type T. That way we can write simpler APIs and even use Scala-like implicit parameters to provide the familiar dotted property access.

**Reply by johnrrowe** [reply]
Thanks for filling me in. I think I agree that it's a niche scenario, so path-dependent types seem like overkill. I was just thinking out loud and exploring how things do or don't translate to MVS is helping me understand it more. My question about linear types in Hylo was more about whether it's a compile error not to consume certain types.

**Comment by dabrahams** [comment]
I'm having trouble understanding this:

> I don't like that this API is inherently unsafe. 

I guess one way to understand it is that the API _can_ be implemented by a model having only safe operations.  But in a generic context, it seems to me it's obviously unsafe, because `project` and `free` can be implemented by unsafe operations.  What am I missing?

**Reply by johnrrowe** [reply]
I didn't phrase that well. What I meant is that there are certain allocators where it would be UB to use a handle created by one instance with another instance. So you either have to make this API unsafe or exclude the allocators where the API would be unsafe. For example

```rust
struct BumpAllocator {
    bytes: Vec<u8>
}

struct BumpHandle<T> {
    offset: usize,
    m: std::marker::PhantomData<T>, // just here because generic param must be used, 0 size
}

impl Allocator for BumpAllocator {
    type Handle<T> = BumpHandle<T>;

    // ignoring alignment
    fn alloc<T, F: FnOnce() -> T>(&mut self, construct: F) -> Self::Handle<T> {
        let final_len = self.bytes.len() + std::mem::size_of::<T>();

        if final_len >= self.bytes.capacity() {
            self.bytes.reserve(final_len - self.bytes.capacity())
        }

        unsafe {
            let object_start = self.bytes.as_mut_ptr().add(final_len);
            std::ptr::write(object_start.cast::<T>(), construct());
            self.bytes.set_len(final_len)
        }

        Handle {
            offset: final_len,
            p: Default::default(),
        }
    }

    unsafe fn project<'s, T>(&'s self, h: &Self::Handle<T>) -> &'s T {
        unsafe { &*self.bytes.as_ptr().add(h.offset).cast::<T>() }
    }

    // Only deallocates memory at the end of the allocator's life, i.e. calling free only drops the object
    unsafe fn free<T>(&mut self, h: Self::Handle<T>) {
        unsafe { self.bytes.as_mut_ptr().add(h.offset).cast::<T>().drop_in_place() }
    }
}
```

A handle created by instance A likely does not refer to a valid object of the same type in instance B, so using `project` and `free` would be unsafe. With extra bookkeeping, you could probably make it safe if you're allowed to throw. With path-dependent types, this usage would be a compile error, making BumpAllocator safe to use even without extra bookkeeping (assuming BumpHandle is a linear type)

**Reply by dabrahams** [reply]
You don't need permission to throw to make something safe, if throwing is possible.  That means there's a suitable dynamic check for the precondition violation, and you can terminate.  Throwing on precondition violations is inappropriate anyway.  Yes, path-dependent types can solve the particular problem that values are paired up inappropriately, but I think there are lots of other potential safety problems in the area of allocators.

**Reply by johnrrowe** [reply]
My fault. I meant something more like terminating or panicking rather than throwing an error that would be handled. I think you can eliminate (almost?) all unsafe client usage with linear, path-dependent types, but there would still be safety concerns when implementing. I think this API breaks down when considering what an inout projection would need to do. If you alloc'd a collection and inside the inout projection you appended to the collection, you might have to realloc, but I don't know how the append function would know what allocator to realloc with.

**Comment by tothambrus11** [comment]
While I still don't have a clear idea how MVS could accommodate allocators, I got to learn about some building blocks that may be useful here:

## Non-escapable values
In Hylo, we can project out views of an object that must not escape from a certain scope (cannot be returned from the function, and we cannot form new whole-part relationships from it (maybe unless that new wrapper is also non-escaping?)).

The problem here is that we cannot project out multiple mutable views from a whole at different times, we must project them out at the same time.

## Linear types, `sink` methods
This was already mentioned above when talking about linear types. `sink` methods allow us to destroy an object. Unless the object is conforming to the `Deinitializable` trait, we must deinitialize a value explicitly using one of the sink methods of the object. Without that, it is impossible to get rid of the object. Such `sink` method may also have required parameters, so we could imagine a `Handle` type with a sink method that requires passing in the allocator:
```hylo
struct Handle<T> {
   fun free(allocator: Allocator<T> inout) sink -> Void {
      allocator.free(...)
   }
}
```

## Remote parts
[Remote parts](https://github.com/hylo-lang/hylo/blob/main/Docs/RemoteParts.md) (or stored projections) are properties that are stored remotely. They are still in the research phase, we haven't yet settled for the design.

## Implicit parameters
Functions will have the option to declare implicit parameters that are automatically supplied to it when a variable declaration before its use site has been declared as `given`:

It may look something like this (syntax not yet finalized):
```hylo
fun do_something(a: Int, allocator: implicit inout Allocator) {
   allocator.allocate(a*2);
}

fun do_something_else(e: Int, allocator: implicit inout Allocator) {
   do_something(a: e) // Allocator parameter is resolved to allocator by type, and supplied implicitly.
}


fun usage() {
   given var alloc = Allocator()

   do_something(a: 5) // Allocator parameter is resolved to alloc by type, and supplied implicitly.
}
```

This may be just a nice sugar to make dependency injection easier. All functions requiring an allocator would need to declare a parameter, but when using implicits, supplying these values at the call sites is done automatically thanks to compile time implicit resolution.

**Comment by dilroop-us** [comment]
I’ve been trying to think about this from the perspective of someone who has used C++ and Rust, and who is also experimenting with designing a small language of my own.

My first instinct was to think about allocators in the usual C++/Rust style: an allocator as something a data structure can store or refer back to when it needs memory. But after reading this discussion and some of the related MVS explanations, I think that model does not really fit Hylo well.

The idea that makes more sense to me is:

> the allocator is not just a service object; it is more like an owning region or container, and the allocated values are parts of that whole.

So instead of allocated values keeping hidden references back to the allocator, maybe the allocator/root should stay involved whenever an operation needs to allocate, grow, move, or destroy storage.

For example, I can imagine a model like this:

```hylo
var arena = BumpArena()

let handle = &arena.allocate(MyValue(...))

&arena.with_inout(handle) { value in
  // temporary access to the value
}
```

Here, the handle is not really a normal reference. It is more like a capability to access a part of the arena. The actual access would happen through the allocator/root, and the projected value would be temporary and non-escapable.

Where I get stuck is with growable collections.

If I have an array, map, string, or some other collection stored inside an arena, and I project it as `inout`, what should happen when I call something like `append` or `insert`? That operation may need to allocate new storage. But if the collection does not store a reference to the allocator, then it does not know where to get memory from.

So I wonder if allocator-backed values need to be split into two categories:

1. Values that can be safely projected and mutated in place because they do not allocate.
2. Values whose operations may allocate, where those operations must go through the allocator/root.

Something like:

```hylo
&arena.append(element, to: arrayHandle)
&arena.insert(key, value, into: mapHandle)
```

This is less like the normal object-method style, but it seems to keep the ownership model clearer. The allocator remains the owner/root, and operations that may allocate do not happen through a hidden allocator reference.

Another possibility could be explicit or implicit allocator parameters for methods like `append`, but I am not sure if that risks bringing back hidden context unless the allocator dependency is very clear in the function type.

My current mental model is:

* handles should probably be linear or non-copyable
* projections should be temporary and non-escapable
* direct mutation is fine when the operation cannot allocate or move storage
* operations that may allocate, grow, relocate, or free should involve the allocator/root explicitly
* arena, pool/slab, and general-purpose allocators may need different APIs instead of one universal allocator model

This may be too early or too simplistic, but I am trying to understand what an allocator model would look like if it fully respects Hylo’s mutable value semantics rather than copying the C++ or Rust model.

Does this distinction make sense?

> Fixed-storage projected values can be mutated directly, but growable or allocating values should route allocation-related operations through the allocator/root.

Or does this still smuggle reference-style semantics back into the design in a way that Hylo should avoid?

**Comment by tothambrus11** [comment]
>Another possibility could be explicit or implicit allocator parameters for methods like append, but I am not sure if that risks bringing back hidden context unless the allocator dependency is very clear in the function type.

That's a great idea, and fair concern. Luckily, implicit parameters always need to be explicitly declared in functions. So you could declare `append` to take an implicit `allocator: inout Allocator` parameter. It is implicit in the sense that the call site doesn't have to manually provide it, but it needs to be present in the local implicit context.

> handles should probably be linear or non-copyable

I think that depends on what they are handles to. We should aim to avoid saving/capturing references, since that decouples them from their root (where they are projected from).

> arena, pool/slab, and general-purpose allocators may need different APIs instead of one universal allocator model

That's interesting, could you recommend some resources to learn more about these, or could you explain why they are different?

> Or does this still smuggle reference-style semantics back into the design in a way that Hylo should avoid?

I think deep memory safety is impossible to achieve because you can always implement a turing machine in safe Hylo and shoot yourself in the foot with it. Turing machines, interpreters, and custom allocators/buffers are examples that implement memory management on a level that the compiler cannot reason about. So we need to essentially rely on correctness of our program to protect ourselves from causing a "memory safety violation" in the simulated/managed memory model. You can see [heartbleed](https://www.heartbleed.com/) as an example of this kind of vulnerability. An art that API designers can aim for is to package up their managed buffers/arenas in a way where upholding Hylo guarantees imply that we also upheld the managed data's guarantees.

In Rust, we can uphold the lifetime relationship between an allocator and its object by constraining the lifetime of the allocated object by the allocator. In Hylo, we can only uphold this relationship if the object is a part of the allocator, or a projection from it. We cannot have independently saved objects referencing the allocator safely without adding runtime checks. We could use something like indices, which would require the allocator to be around whenever we want to access objects. It doesn't guarantee that we access the object from the right allocator, but we can either say that's not an issue or add extra runtime checks (see https://gist.github.com/tothambrus11/63bece9efba1fb6162f03abc2391b1bf)

**Reply by dabrahams** [reply]
There's no such thing as a memory safety violation in a Turing machine AFAICT, because it is fully dynamically typed. What does "deep memory safety" mean?

**Reply by tothambrus11** [reply]
You are right, the Turing machine is not the right example here.

The heartbleed vulnerability allowed an attacker to read parts of a long-lived buffer that it wasn't supposed to read except the part of it that was allocated for the request. This was caused by a missing bounds check in an abstraction that was supposed to ensure certain guarantees. A memory safe language wouldn't have prevented this issue. I now realize the main problem is adding an abstraction without ensuring the abstraction's invariants are upheld. Then operations on that abstraction operate on possibly broken preconditions. This is analogous to a programming language being an abstraction over assembly - if the invariants of the language are ensured, it's a safe language. But a language doesn't have to be a special place to stop abstracting, so the care for ensuring invariants of new abstractions should also not stop there. 

A SmallVector with fixed-sized capacity can be implemented using a `type SmallVector<T> = { T[8], size: UInt8 }`.  Any element of it could be accessed safely according to the language. But it would be nice if the users were prevented to access/modify elements outside of `[0, size)`, so the naive accessor could be called small-vector-invariant-unsafe.

So what I meant by deep safety is the property of a program where all abstractions enforce their invariants.

If we had an `unsafe` keyword in the language, I would annotate all methods `unsafe` that don't enforce the type's invariants/the function's preconditions.

**Reply by dabrahams** [reply]
Then you'd be abusing the keyword. Upholding invariants does not compose and is therefore not a safety property.  Furthermore, in Hylo, `unsafe` means exactly one thing: _can cause undefined behavior if used incorrectly_.
